File: backend-spring/src/main/java/com/codequest/security/FirebaseTokenFilter.java
package com.codequest.security;

import com.google.firebase.auth.FirebaseAuth;
import com.google.firebase.auth.FirebaseAuthException;
import jakarta.servlet.FilterChain;
import jakarta.servlet.ServletException;
import jakarta.servlet.http.HttpServletRequest;
import jakarta.servlet.http.HttpServletResponse;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.security.authentication.UsernamePasswordAuthenticationToken;
import org.springframework.security.core.context.SecurityContextHolder;
import org.springframework.stereotype.Component;
import org.springframework.web.filter.OncePerRequestFilter;
import java.io.IOException;
import java.util.Collections;

@Component
public class FirebaseTokenFilter extends OncePerRequestFilter {

    private static final Logger logger = LoggerFactory.getLogger(FirebaseTokenFilter.class);
    private final FirebaseAuth firebaseAuth;

    public FirebaseTokenFilter(FirebaseAuth firebaseAuth) {
        this.firebaseAuth = firebaseAuth;
    }

    @Override
    protected void doFilterInternal(HttpServletRequest request, HttpServletResponse response, 
                                  FilterChain filterChain) throws ServletException, IOException {
        
        String path = request.getRequestURI();
        String method = request.getMethod();
        
        // Это лог появится ПЕРВЫМ в консоли бэкенда
        logger.info(">>> [API REQUEST] {} {}", method, path);

        try {
            String token = extractToken(request);

            if (token != null) {
                try {
                    var decodedToken = firebaseAuth.verifyIdToken(token);
                    String uid = decodedToken.getUid();
                    
                    // Устанавливаем в ваш кастомный контекст
                    SecurityContext.setUserId(uid);
                    
                    // Устанавливаем в стандартный Spring Security контекст
                    UsernamePasswordAuthenticationToken authentication = new UsernamePasswordAuthenticationToken(
                            uid, null, Collections.emptyList());
                    SecurityContextHolder.getContext().setAuthentication(authentication);
                    
                    logger.info(">>> User {} authenticated for path {}", uid, path);
                } catch (FirebaseAuthException e) {
                    logger.error(">>> Firebase Auth Error for {}: {}", path, e.getMessage());
                    response.setStatus(HttpServletResponse.SC_UNAUTHORIZED);
                    response.getWriter().write("Invalid Firebase token");
                    return;
                }
            } else {
                logger.warn(">>> No Bearer token found for path: {}", path);
            }

            filterChain.doFilter(request, response);
        } finally {
            // Очищаем оба контекста после завершения запроса
            SecurityContext.clear();
            SecurityContextHolder.clearContext();
        }
    }

    private String extractToken(HttpServletRequest request) {
        String authHeader = request.getHeader("Authorization");
        if (authHeader != null && authHeader.startsWith("Bearer ")) {
            return authHeader.substring(7);
        }
        return null;
    }

    @Override
    protected boolean shouldNotFilter(HttpServletRequest request) {
        String path = request.getRequestURI();
        // Разрешаем проходить без проверки токена для этих путей
        return path.contains("/health") || 
               path.contains("/swagger") || 
               path.contains("/v3/api-docs") || 
               path.contains("/docs");
    }
}

File: backend-spring/src/main/kotlin/com/codequest/security/FirebaseTokenFilter.kt
package com.codequest.security

import com.google.firebase.auth.FirebaseAuth
import com.google.firebase.auth.FirebaseAuthException
import io.github.microutils.kotlin.logging.KotlinLogging
import jakarta.servlet.FilterChain
import jakarta.servlet.http.HttpServletRequest
import jakarta.servlet.http.HttpServletResponse
import org.springframework.stereotype.Component
import org.springframework.web.filter.OncePerRequestFilter

private val logger = KotlinLogging.logger {}

@Component
class FirebaseTokenFilter(
    private val firebaseAuth: FirebaseAuth
) : OncePerRequestFilter() {

    override fun doFilterInternal(
        request: HttpServletRequest,
        response: HttpServletResponse,
        filterChain: FilterChain
    ) {
        try {
            val token = extractToken(request)

            if (token != null) {
                try {
                    val decodedToken = firebaseAuth.verifyIdToken(token)
                    val uid = decodedToken.uid
                    SecurityContext.setUserId(uid)
                    logger.debug { "User authenticated: $uid" }
                } catch (e: FirebaseAuthException) {
                    logger.warn { "Invalid Firebase token: ${e.message}" }
                    response.sendError(HttpServletResponse.SC_UNAUTHORIZED, "Invalid token")
                    return
                }
            }

            filterChain.doFilter(request, response)
        } finally {
            SecurityContext.clear()
        }
    }

    private fun extractToken(request: HttpServletRequest): String? {
        val authHeader = request.getHeader("Authorization") ?: return null
        return if (authHeader.startsWith("Bearer ")) {
            authHeader.substring(7)
        } else {
            null
        }
    }

    override fun shouldNotFilter(request: HttpServletRequest): Boolean {
        val requestUri = request.requestURI
        return requestUri.contains("swagger") ||
                requestUri.contains("docs") ||
                requestUri == "/health"
    }
}

